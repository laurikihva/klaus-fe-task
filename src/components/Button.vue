<template>
  <button :class="computedClasses" type="button" :disabled="disabled">
    <span class="flex items-center gap-2">
      <Icon v-if="icon" :name="icon" />
      {{ label }}
    </span>
  </button>
</template>

<script lang="ts" setup>
import { toRefs, computed } from "vue";

import Icon, { TIconName } from "./Icon.vue";

export type TButtonColor = "primary" | "secondary";
export interface IButtonProps {
  label?: string;
  color?: TButtonColor;
  icon?: TIconName;
  small?: boolean;
  disabled?: boolean;
}

const props = withDefaults(defineProps<IButtonProps>(), {
  color: "primary",
});
const { color, small, disabled } = toRefs(props);

const buttonColor = computed(() => {
  return {
    primary: "text-white bg-blue-400 enabled:hover:bg-blue-500",
    secondary: "text-gray-500 bg-white border border-gray-400 enabled:hover:bg-gray-400",
  }[color.value];
});

const computedClasses = computed(() => {
  const base = "rounded text-sm tracking-wider font-medium";
  const isSmall = small.value ? "py-1 px-2 min-h-[32px]" : "py-2 px-3 min-h-[40px]";
  const isDisabled = disabled.value ? "cursor-not-allowed opacity-85" : "";
  return [base, buttonColor.value, isSmall, isDisabled];
});
</script>
