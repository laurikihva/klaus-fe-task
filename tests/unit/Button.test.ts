import { mount } from "@vue/test-utils";

import Button from "@/components/Button.vue";

const wrapper = mount(Button, {
  props: {
    label: "My button",
    icon: "trash",
    small: true,
    disabled: true,
  },
});

describe("Loads and displays Button component", () => {
  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("contains label and base classes", () => {
    expect(wrapper.text()).toContain("My button");
    expect(wrapper.find("svg")).toBeTruthy();
    expect(wrapper.classes("rounded")).toBe(true);
    expect(wrapper.classes("cursor-not-allowed")).toBe(true);
  });
});
