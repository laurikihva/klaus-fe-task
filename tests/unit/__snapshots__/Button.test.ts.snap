// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loads and displays Button component > renders correctly 1`] = `"<button class="rounded text-sm tracking-wider font-medium focus:shadow-primary focus:outline-none text-white bg-blue-400 enabled:hover:bg-blue-500 py-1 px-2 min-h-[32px] cursor-not-allowed opacity-85" type="button" disabled=""><span class="flex items-center gap-2"><svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 18 18" class="w-[14px] h-[14px]"><path d="M15.75 3h-2.325A3.76 3.76 0 0 0 9.75 0h-1.5a3.76 3.76 0 0 0-3.675 3H2.25a.75.75 0 0 0 0 1.5H3v9.75A3.754 3.754 0 0 0 6.75 18h4.5A3.754 3.754 0 0 0 15 14.25V4.5h.75a.75.75 0 1 0 0-1.5m-7.5-1.5h1.5A2.255 2.255 0 0 1 11.872 3H6.128A2.255 2.255 0 0 1 8.25 1.5m5.25 12.75a2.25 2.25 0 0 1-2.25 2.25h-4.5a2.25 2.25 0 0 1-2.25-2.25V4.5h9z"></path><path d="M7.5 13.5a.75.75 0 0 0 .75-.75v-4.5a.75.75 0 0 0-1.5 0v4.5a.75.75 0 0 0 .75.75m3 0a.75.75 0 0 0 .75-.75v-4.5a.75.75 0 1 0-1.5 0v4.5a.75.75 0 0 0 .75.75"></path></svg> My button</span></button>"`;
